### ISSUE TITLE
PWM sound pretty different from the real thing.

### ISSUE THREAD

**Issue by oliverschmidt** [AppleWin/AppleWin#1088]
With the recently added Uthernet II emulation, my program A2Stream (https://github.com/oliverschmidt/A2Stream) runs in general fine on AppleWin - thanks :-)

However, the actual sound output is pretty different from the real thing. Please note that I'm not saying, that anything is technically wrong. But as always in emulation, the real thing is the benchmark, isn't it? So whatever makes AppleWin sound so different, should from my POV be changed/fixed/improved/[you name it].

[A2Stream.zip](https://github.com/AppleWin/AppleWin/files/8496683/A2Stream.zip) contains four .mp3 files:

- IIgs.mp3 was created by connecting the IIgs headphone jack to the microphone jack of a Windows PC.
- AppleWin.mp3 was created via loopback of the Windows audio output while running AppleWin.

While you may argue that the way I created the .mp3's might have introduced some artifacts, I can assure you that they both sound exactly as the human ear perceives the original sound that was recorded.

For the sake of elucidation, I extracted a short section of silence from the two .mp3's above:
- Silence-IIgs.mp3
- Silence-AppleWin.mp3

Some technical background regarding A2Stream: It uses a deterministic flow of 46 machine cycle code artifacts to generate a 22.050 kHz PWM signal - exactly like DAC522 does.

**Comment by tomcw** [comment]
Thanks Oliver for raising this issue.

I'm going to try to repro this on my local setup - but it'll be a //e not a IIgs for the real system.

**Comment by oliverschmidt** [comment]
> I'm going to try to repro this on my local setup

Thanks :-)

> but it'll be a //e not a IIgs for the real system

The enhanced //e with 80 col. board is the primary target system. Adding a file OUTPUT.TAPE (see https://github.com/oliverschmidt/A2Stream/blob/main/README.md) redirects the sound output to the cassette/tape out jack ($C020). Adding a suitable device allows to assess the sound quality. It is supposed to be identical to the IIgs.mp3 I posted above.

**Comment by tomcw** [comment]
Thanks for the info. I managed to do a few recordings. 

The real //e sounds clearer, whereas the AppleWin output has a "background noise" (hiss) component.

---

As an aside (and more a reference/note to myself), all recordings suffered from sporadic, occasional clicking...
[A2e-slot2-outputTape-lineIn-TimeToSayGoodbye(1)-clicky.zip](https://github.com/AppleWin/AppleWin/files/8505752/A2e-slot2-outputTape-lineIn-TimeToSayGoodbye.1.-clicky.zip)
(eg. playback from 1m:16s + they are easy to spot in Audacity)

Also with OUTPUT.TAPE enabled (ie. output to $C020, and not $C030) I would get occasional clicks from my //e speaker! These also correlated with clicks occurring in my recording.

But when listening to normal $C030/speaker playback (ie. OUTPUT.TAPE disabled), then I couldn't perceive any extra clicks.

Anyway, probably not important in the context of this issue. I will just assume that my setup is not very good, and suffers from some noisy interference somewhere.

**Comment by oliverschmidt** [comment]
> The real //e sounds clearer, whereas the AppleWin output has a "background noise" (hiss) component.

That's exactly as I perceive it. I couldn't describe it better. And as I wrote before: On the one hand, it's great that AppleWin is able to reproduce that audio output at all in a consistent way. But on the other hand, it's a pity that there's this "background noise" keeping one from actually enjoying the experience - and keeping one from fully understanding, how well A2Stream actually sounds ;-)

Apart from the clicks/distortions you mention, the `A2e-slot2-outputTape-lineIn-TimeToSayGoodbye(1)-clicky.mp3` seems to sound very much like what I get from my IIgs - which is great to know for me - thanks for sharing :-)

And finally regarding those clicks/distortions: I'm almost sure that they come from an imperfect electric connection - likely at the tape out jack on your //e, as that jack most certainly wasn't used for decades. The reason for me thinking so is, that there's some relation between the loudness of the song and the amount of distortions, which is typical for such issues. Using some contact spray and turning the plug in the jack left and right > 20 times is many/most times enough to fix the issue.

**Comment by audetto** [comment]
Attached the linux version

[Spkr.zip](https://github.com/AppleWin/AppleWin/files/8580476/Spkr.zip)

I generated with `RIFF_SPKR`, so it might miss some artefacts.

To me it highlights some issues (linux only?)
The wave is not symmetric at all, not centred around 0, but I still dont master the speaker code enough to understand what causes it.

**Comment by sicklittlemonkey** [comment]
@oliverschmidt 
Have you tried it on other emulators to compare how they handle it?
In particular I'm thinking of Catakig and OpenEmulator for the Mac.
Catakig's author, Colin, got Michael Mahon's help implementing an FIR filter.
https://github.com/aufflick/Catakig/blob/master/Source/LibAppleII/A2Computer/Audio.m
(There are a couple of forks, but it looks like you'd need to build it yourself.)

**Comment by oliverschmidt** [comment]
* As far as I know, the mentioned emulators don't emulate the Uthernet II card.
* I don't own a Mac.

**Comment by oliverschmidt** [comment]
> Attached the linux version [...]

I consider this pretty interesting:

Windows AppleWin produces a certain (from my POV rather disturbing) background noise with a distinct frequency. I'm not good at such things, but it might be around 100Hz.

That distinct noise it missing completely from Linux AppleWin - at least as far as I can tell. However, it seems to produce more pink noise.

So least to me, it seems that this distinct, disturbing Windows AppleWin noise is introduced by code not shared with Linux AppleWin.

BTW: All in all, if I had to choose between the Windows and the Linux audio, I'd choose the Linux one.

**Comment by tomcw** [comment]
>I generated with `RIFF_SPKR`

btw, not in a release build yet, but at 08c730c6473ec4ebd5f38be4d5c78b4e9502cb76 I added new command line options to capture these RIFF/WAV files directly from the cmd line using `-wav-speaker file`.

**Comment by sicklittlemonkey** [comment]
> * As far as I know, the mentioned emulators don't emulate the Uthernet II card.

No, but presumably you could load a sample into RAM and have your routine play a few seconds.
It should be enough to compare the sound fidelity of different open source emulators.

> * I don't own a Mac.

Neither do I. Perhaps someone could run the test for you.
Even comparing some Windows emulators might be helpful if the sound fidelity is better.

**Comment by oliverschmidt** [comment]
> Have you tried it on other emulators to compare how they handle it?

Sorry for interpreting `it` as the `the program` ;-))

> No, but presumably you could load a sample into RAM and have your routine play a few seconds.

Now I see. That would be a major effort - presumable several magnitudes above your expectation. I don't see me doing that.

**Comment by sicklittlemonkey** [comment]
@oliverschmidt 
> Now I see. That would be a major effort - presumable several magnitudes above your expectation. I don't see me doing that.

Fair enough. But it uses exactly the same technique as DAC522?
So does something like RTSynth suffer similar problems in AppleWin compared to a real Apple II?
http://michaeljmahon.com/RTSynth.html

**Comment by oliverschmidt** [comment]
> So does something like RTSynth suffer similar problems in AppleWin compared to a real Apple II?

I'd say 'yes'.

**Comment by genewitch** [comment]
there's quantization or digital artifacting in the applewin sample. No idea how it was programmed, but it sounds like some extreme compression was used/applied as a filter. A lot of my old samples that i recorded on things that didn't export .wav files (which is like bmp, it's basically raw data) sound like this, now, on hi-fi soundcards and headphones.

you can hear a bit of it in the non applewin sample as well. maybe generate a perfect sine wave at 1khz (or 500hz) and see where the artifacts are. Everything is sine waves, after all, so that will show you everything in the pipeline that messes with the sine wave.

**Comment by tomcw** [comment]
@genewitch - thanks for the suggestion (1KHz sine wave). I need to find some time, but it seems like a good thing to try.

**Comment by tomcw** [comment]
No real progress, but let me just document some things I've noticed & tried so far.
NB. This is mostly a note to myself as to what I've done, and is not intended as a formal (or even authoritative/precise) note!

Two screenshots from AppleWin, using `-wav-speaker file` to grab the raw PCM data that AppleWin is generating.
(Two independent runs streaming "Time To Say Goodbye")

Notice that in both cases, that one half (1st bottom half, 2nd top half) is pretty constant, ie. flat:
IE. alternate samples are mostly constant.

![AppleWin-wav-speaker-TimeToSayGoodbye](https://user-images.githubusercontent.com/6696896/176997821-be0b9feb-3acc-4fc4-9bc2-6a391a670c63.PNG)

![AppleWin-wav-speaker-TimeToSayGoodbye2](https://user-images.githubusercontent.com/6696896/176997822-e4e5f33f-8438-4984-a387-0eee86b90317.PNG)

I looked into the data and think it's just an aliasing effect.

In very rough terms, AppleWin is "sampling" the 1MHz accesses to $C030 every 23 cycles (=1M/44100), so ~44.1kHz (ie. take the mean speaker value each 23 cycles). And the A2Stream does PWM at 22050Hz sample rate (46 cycles.) Accesses to $C030 appear to be 23cycle+/-n (where n is small).

Since $C030 accesses can occur <23 samples then 2 accesses can occur in the same 23-cycle interval.
Here's an extract from some AppleWin logging showing 2 accesses in 23 cycles:
```
Mean=27068, Total=622573: -32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, 32767, -32767, 
```

My knowledge of audio theory is poor - perhaps I need to "sample" at 2x 44.1kHz (eg. Nyquist).

btw, as a quick experiment, I changed the frequency from 44.1kHz to 48kHz, which helps a bit with the above, but produces audible (and **visible** from the screenshot) "ringing" - so a different aliasing effect:
(and doesn't help with the underlying noise/hiss)

![AppleWin-wav-speaker-TimeToSayGoodbye3-48000Hz](https://user-images.githubusercontent.com/6696896/176998473-ad8b70e3-b3d3-4579-8cf7-fec7d899f7a5.PNG)

Still to try: the 1KHz sine wave suggested above.

**Comment by oliverschmidt** [comment]
First of all, thanks for your efforts :-)

Unfortunately, I can't add useful info. The only thing that comes into my mind is, that the difference between the first and second screenshot almost certainly comes from the speaker happening to be either 'on' or 'off by the time A2Stream starts to trigger it.

**Comment by audetto** [comment]
One thing I wanted to do is to decouple the problem of audio generation from the emulator.

I mean, if we extract the cycle counters at which the speaker is accessed, can anyone produce a good wav file offline? Using python for instance.

Then, we must ensure emulation has as little impact as possible.

Or maybe this is obvious to everybody already?

**Comment by oliverschmidt** [comment]
There is no need to extract the cycle counters, at which the speaker is accessed. I can easily provide them here, as the whole A2Stream code is built around the perfect control of that very aspect.

As mentioned, there's a 46 machine cycle loop (aka PWM pulse rate). If we number those machine cycles from 0 to 45, we have:

* The first access starting the PWM duty cycle is always at machine cycle 0.
* The second access ending the PWM duty cycle varies between machine cycle 4 and 39. So there are 36 different duty cycles.

Any PCM audio data scaled down to sample values between 0 and 35 can be used as test data. But if one feels, that tests needs to be done with actual A2Stream data (which I personally do NOT), then https://github.com/oliverschmidt/A2Stream/blob/main/gena2stream.c#L62 ff. can be used as a guidance on how to retrieve those [0-35] sample values from an .a2stream file.

Note: As discussed, DAC522 is very similar. Actually, it's a subset of A2Stream: Its second speaker access varies between machine cycle 6 and 37, resulting 32 different duty cycles (aka 5 bit PWM resolution).

**Comment by sicklittlemonkey** [comment]
@audetto 
> One thing I wanted to do is to decouple the problem of audio generation from the emulator.

I think that's a great idea.

My life is a mess, so all I can contribute is the following:
https://groups.google.com/g/comp.sys.apple2/c/i9e1s_BAJ20/m/NkRtSdQzrV4J

The thread is back in 2005, when I discovered that Michael Mahon was the DSP guru who helped Colin Klipsch write the FIR filter that he used in Catakig. Michael discusses some other sound emulation details in that thread, and Linards Ticmanis, who also contributed to AppleWin, describes a simple "decimation" scheme that he used in YAE.

Then again, other emulators, JACE comes to mind, sounded good without anything fancy.

Cheers,
Nick.

**Comment by sh95014** [comment]
@oliverschmidt I thought I'd try a2stream on a Mac to see how it works with my new CoreAudio-based sound output/recording (not expecting very different results from Linux except with latency), but I can't seem to get past "Connecting to port 80":

![Mariani Screenshot](https://user-images.githubusercontent.com/95387068/180573185-fd54dda0-cb44-44f5-8711-47626d06d904.png)

The Mac version inherits Uthernet support via libslirp from Linux, and it does manage to get NTP time with the ip65 test app, so I'm not sure what's up.

**Comment by Michaelangel007** [comment]
@tomcw I forgot to file this bug a few years back when I was working on _Nox Archaist._  This year's Kansasfest reminded me that I should probably file this bug but I see it has already been done.

If you haven't watched it already, Kriss Kennaway had an excellent [talk today](https://youtu.be/Uf-JvoZxnso?t=5558) at this year's Kansasfest (1:32:38) explaining in clear detail how he implemented 5-bit PWM and Delta Modulation to get another bit of accuracy.

Oliver is actually incorrect when he writes _"Please note that I'm not saying, that anything is technically wrong."_

Audio IS wrong in AppleWin.  Normal audio code doesn't drive the speaker fast enough which is why for 99% of audio apps you will never notice a problem.  Kriss is using 5-bit PWM at a variable frequency > 22 KHz which _does_ drive the speaker fast enough which showcases AppleWin's emulation is incorrect. Running _Nox Archaist_ on a real Apple //e has silence where there is a _very faint_ high pitched whine in AppleWin.

> The real //e sounds clearer, whereas the AppleWin output has a "background noise" (hiss) component.

That's a very good summary.

I'll attach the two files that Kris sent me back in 2020:

* Audio recording of Nox Archaist in AppleWin: [20200917-171934.zip](https://github.com/AppleWin/AppleWin/files/9172918/20200917-171934.zip)

* Old Beta build 0.82 of NA which demoes the high pitched whine:  [Nox Archaist Beta (v0872-beta).zip](https://github.com/AppleWin/AppleWin/files/9172921/Nox.Archaist.Beta.v0872-beta.zip)

Sorry for not filing this sooner.  Was extremely busy with Real Life TM.

**Comment by oliverschmidt** [comment]
> Oliver is actually incorrect when he writes "Please note that I'm not saying, that anything is technically wrong."

I did not state, that I say that everything is correct. I did state, that I don't say anything. From that perspective, I can't be wrong.

**Comment by audetto** [comment]
> The Mac version inherits Uthernet support via libslirp from Linux, and it does manage to get NTP time with the ip65 test app, so I'm not sure what's up.

It is 2 different things.

A2Stream uses TCP sockets which are implemented via OS sockets.
They have been tested on Windows and on Linux where there were minor differences.

Someone (was it you?) tested them on MacOS, but it was a while ago.
See for instance:

https://github.com/AppleWin/AppleWin/blob/c9f5df1e1eda992282a3a71f67237d3f245be884/source/Uthernet2.cpp#L274
https://github.com/AppleWin/AppleWin/blob/c9f5df1e1eda992282a3a71f67237d3f245be884/source/Uthernet2.cpp#L706
https://github.com/AppleWin/AppleWin/blob/c9f5df1e1eda992282a3a71f67237d3f245be884/source/Uthernet2.cpp#L814

and enable some logging, probably these 3: https://github.com/AppleWin/AppleWin/blob/c9f5df1e1eda992282a3a71f67237d3f245be884/source/Uthernet2.cpp#L92-L94

**Comment by sh95014** [comment]
Thanks! Turns out my problem is that `SOCK_NONBLOCK` is a Linuxism that causes the socket to fail to open on macOS. I'll submit a PR.

Once that's fixed, there seems to be a bug on my CoreAudio version (that I'll go hunt down separately), but once worked around the audio seems a lot cleaner than the AppleWin.mp3 clip. The macOS+SDL version also seems clean.

**Comment by audetto** [comment]
The ifdef for sock_nonblock has been added after another  macos user proposed it. Will try to find out

**Comment by sh95014** [comment]
> https://github.com/audetto/AppleWin/issues/39#issuecomment-944988587

The `O_NONBLOCK` workaround still hangs, unfortunately. I found this to work instead:

```
@@ -916,7 +916,7 @@ void Uthernet2::resetRXTXBuffers(const size_t i)
 void Uthernet2::openSystemSocket(const size_t i, const int type, const int protocol, const int status)
 {
     Socket &s = mySockets[i];
-#ifdef _MSC_VER
+#if defined(_MSC_VER) || defined(__APPLE__)
     const Socket::socket_t fd = socket(AF_INET, type, protocol);
 #else
     const Socket::socket_t fd = socket(AF_INET, type | SOCK_NONBLOCK, protocol);
```

**Comment by sh95014** [comment]
One more before I give up...

The input LPCM stream seems to be substantially bigger than it claims to be. I timed the CoreAudio callback and it pretty consistently asks for ~44,218 frames per second, but that doesn't manage to empty the `IDirectSoundBuffer`. I tried forcing the buffer to empty on each call to `Read()`:

```
-  this->myPlayPosition = (this->myPlayPosition + dwReadBytes) % this->mySoundBuffer.size();
+  this->myPlayPosition = (this->myPlayPosition + available) % this->mySoundBuffer.size();
```

and that seems to make the CoreAudio implementation happy except the sound is tinny and noisy, which is expected because `available` - `dwReadBytes` is typically ~1,800 bytes.

So I think this supports @audetto 's:

> decouple the problem of audio generation from the emulator

idea.

**Comment by audetto** [comment]
I think one of the issues in linux is that emulation runs a frame at a time. In windows is 1ms. This means that the audio code needs to handle a bigger variability of data in buffer. Imagine what happens if the code runs just before or just after the buffer is emptied of 16ms. 

https://github.com/AppleWin/AppleWin/blob/c9f5df1e1eda992282a3a71f67237d3f245be884/source/Speaker.cpp#L748

This code probably constantly over / under shoots the right amount.

I ve been wanting to add diagnostics to see if this is correct, but never got to do it.

If we move to a callback based approach (in linux), we could probably shortcut the whole thing, but this is entirely to be seen.

**Comment by univta0001** [comment]
The AppleWin version may not be generating as faithful the original Apple 2. I might be wrong (as i don't have the real Apple nowadays), but the following code should generate an almost inaudible tone.

300: LDA $C030
303: JMP 300

AppleWin generates a very high pitch tone. It should be silent. If you use MAME or JACE, the tone is silent.

In MAME, the audio samples are FIR filtered before down-sampling. JACE is doing average filtering if i am not wrong.  Another way to make the speaker output more "correct" is to simulate the speaker frequency response as mentioned by Kris Kennaway in KansasFest 2022 (https://www.kansasfest.org/wp-content/uploads/2022/08/KFest2022-Kennaway-a2-audio.pdf)

In short, the speaker output y is given by the following formula

 y[t] = c1 * y[t-2]  - c2 *  y[t-1] + x[t]

where x[t] is the raw samples at each cycle, c1 is dependent on the natural frequency of the speaker, and c2 is the damping factor. In the Kansasfest 2022 paper, the natural  frequency of the speaker is measured to be 3880Hz. 

Recently i have implemented this, and the A2Stream and RT.Synth sounds better but there will be slight slowdown in the emulator as the audio samples need to be filtered for each clock cycle. Accurapple is also exploring to implement this (https://gitlab.com/wiz21/accurapple/-/blob/composite/speaker/snd_fit.py)

**Comment by tomcw** [comment]
>Recently i have implemented this, and the A2Stream and RT.Synth sounds better

Great! Can you attach some A2Stream .wav examples so that we can hear the difference between current AppleWin, and using this formula?

btw, since (in A2Stream) the raw sample only changes every 23+/-n (for small n) cycles, then presumably you can do the y[t] calc. as a just-in-time update... ie. whenever the speaker is toggled, and calc. all the intermediate y[t] values.

I'm just considering a (premature) optimisation here so that there's not a requirement to do a calc every cycle.

**Comment by oliverschmidt** [comment]
>  (in A2Stream) the raw sample only changes every 23+/-n (for small n) cycles

That `n` can be as big as 18. See https://github.com/AppleWin/AppleWin/issues/1088#issuecomment-1173032565.

**Comment by univta0001** [comment]
Hi, in attaching the audio samples of Darude Sandstorm of 20 seconds from applewin and from my emulator using the formula. The Darude sandstorm a2stream stream can be found at http://a2retro.de/a2s/Sandstorm.a2stream. 

The audio samples are captured in Audacity and save in raw format

1) 44100 bits stereo and 32-bit float

[Uploading audio_samples_darude.zip…]()

**Comment by oliverschmidt** [comment]
A few comments from my side:
1. The FIR audio seems to have some issue maybe related to higher volumes. It's for the first time audible at ~18.75. Then it happens often from ~25.50 on.
2. The Emu6502 audio seems like having gone through a pretty strong low pass filter compared to the AppleWin audio. Is this deliberate because it sounds more like the physical speaker of an Apple II?
3. I'm __*really*__ astonished how hard it obviously is to get this right! I mean, my IIgs allows me to get the audio from the headphone jack with a very nice frequency response but yet without any of those artifacts we're talking here about. And that without __*any*__ fancy preprocessing of the audio data before playback.

**Comment by univta0001** [comment]
Hi thanks for the note.

1. The issues found in ~18.75 and ~25.50 is attributed to the audio buffering issues in my emulator. Attached is the corrected copy. ([emu6502_darude_fir_44100_stereo_32float_corrected.zip](https://github.com/AppleWin/AppleWin/files/10974857/emu6502_darude_fir_44100_stereo_32float_corrected.zip))

2. Yes, I agree with you that the audio on Kansasfest is strongly low pass. The algorithm mentioned above is to model the physical characteristics of the speaker in the Apple ][.

3. The complexity is due to the aliasing of the high frequency components in the audio as low frequency. Anti-alias filter would need to be applied before decimation of the audio at 23 cycles to get 44.1 kHz. The challenge is in the finding the proper anti-alias filter.

**Comment by oliverschmidt** [comment]
1. Thanks for the corrected audio!
2. I personally don't like this. But I also don't like the emulation of 60Hz flicker and CRT distortion so my taste is probably weird ;-)
3. Thanks for the explanation :-)

**Comment by tomcw** [comment]
Comparing the 3 (AppleWin vs Emu6502 using Kansasfest 2022 algorithm vs Emu6502 using FIR filter), you can visually see that the the two Emu6502 samples remove the aliasing that AppleWin has.

NB. The middle one (Emu6502 using Kansasfest 2022 algorithm) is inverted (y-axis) compared to the other two. (This makes no difference to the sound.)

![Compare3_darude_44100_mono_32bit_float](https://user-images.githubusercontent.com/6696896/225459973-1ac2ab8e-f6df-414e-ac16-0eff3651574b.PNG)

**Comment by audetto** [comment]
> Hi, in attaching the audio samples of Darude Sandstorm of 20 seconds from applewin and from my emulator (emu6502) using the formula.

Do you mind sharing a link to the speaker implementation of emu6502.

**Comment by audetto** [comment]
> In short, the speaker output y is given by the following formula
> 
> y[t] = c1 * y[t-2] - c2 * y[t-1] + x[t]
> 
> where x[t] is the raw samples at each cycle, c1 is dependent on the natural frequency of the speaker, and c2 is the damping factor. In the Kansasfest 2022 paper, the natural frequency of the speaker is measured to be 3880Hz.
> 
> Recently i have implemented this, and the A2Stream and RT.Synth sounds better but there will be slight slowdown in the emulator as the audio samples need to be filtered for each clock cycle. Accurapple is also exploring to implement this (https://gitlab.com/wiz21/accurapple/-/blob/composite/speaker/snd_fit.py)

I have tried to implement the above formula (from snd_fit.py) and I get the following coefficients

```
c1 = 1.996989416488857
c2 = 0.997582925839347
```

but if I feed `x[t]` = 1, I get these numbers as output: 0, 1, 2.99, 5.98, 9.96 ,,, and they grow and grow until some -1 arrives, but the whole thing explodes.

How is this supposed to work?

**Comment by audetto** [comment]
Maybe I am misunderstanding the meaning of `x[t]`, I though it would be the level of the square wave at 1MHz that can jump from -1 to 1.
I tried as well with `x[t]` as the change of the above wave, but the output of the filter goes to levels which at 10x, I can truncate it, but it just produces noise.

**Comment by univta0001** [comment]
Hi, the absolute value of x[t] must be within 0 to 1. In emu6502 implementation, y[t] is normalized by dividing it to 4000. This value is experimentally determined. The audio code of emu6502 is implemented in Rust. Your c1 and c2 is different from emu6502 but it should work. 

Here is the complete code.

    fn filter_response(&mut self, value: Channel) -> f32 {
        let c1 = 1.9955211;
        let c2 = 0.996093;

        let y = c1 * self.filter_tap[0] - c2 * self.filter_tap[1] + (value as f32) / 32768.0;
        self.filter_tap[1] = self.filter_tap[0];
        self.filter_tap[0] = y;

        let mut return_value = y / 4000.0;
        if return_value < -1.0 {
            return_value = -1.0;
        } else if return_value > 1.0 {
            return_value = 1.0;
        }

        return_value
    }

**Comment by audetto** [comment]
This makes more sense.
I tried to locate the source code of emu6502 but I have not found it. Is it available?

**Comment by univta0001** [comment]
It is not available in public. It is private github repository.

**Comment by audetto** [comment]
Would you be able to share some the downsampling algorithm you have implemented?

**Comment by univta0001** [comment]
In emu6502, the filter response function is called in every clock tick. The audio playback is 48000 Hz. Therefore the audio buffer is filled with the data from filter response function roughly every 21 ticks ( 1 MHz / 48000 Hz ). There is no interpolation or low pass being employed.

**Comment by univta0001** [comment]
Attached the source code changes to Speaker.cpp ([Speaker.zip](https://github.com/AppleWin/AppleWin/files/12251374/Speaker.zip)). The source code is based on the description above. As i am not totally familiar with the sound code in AppleWin, there are still some sound glitches. Initial testing with it, playing TimeToSayGoodbye sounds better.

**Comment by wiz21b** [comment]
Dear All,

Looks like I'm not the only one on audio :) I've been toying with sound emulation since several months too. I have implemented something similar to Kris Kenaways routine. I have a fully working implementation with all the explanations of the maths. I also do the downsampling (almost) correctly and on top of that I'm doing a full FFT equalization (realtime of course). Hopefully I'll get as close to a //e as I can (if I can do a proper recorrding of the //e sound). If you're interested, I can show the code (rust)  and the maths.

**Comment by tomcw** [comment]
@univta0001 - thanks for attaching your code. I don't have time at the moment to test / look at it, but I fully intend to and hopefully we can improve on the existing AppleWin implementation.

**Comment by tomcw** [comment]
@wiz21b - thanks for chipping in. Yes, having another implementation (even in Rust), but also with the maths would be very useful. Any pooling of knowledge on this topic will hopefully improve the  quality overall

**Comment by tomcw** [comment]
@wiz21b... so please share your code too. Thanks.

**Comment by wiz21b** [comment]
Hi! I will prepare some maths as soon as i come back from holidays in about ten days.  I'll try to get a simpler Way to compute the C1 ans C2 constants.

**Comment by wiz21b** [comment]
While on the beach I drafted the initial idea of the recurrence expression, C1 and c2 are built on the basis of the code below. It's not complete yet. I'll do the complete solution once back home and will typeset the maths too. I'm fairly confident I have all the pieces now

```

dt = pi / 1000
x1 = sin(0)
x2 = sin(dt)
for i in range(2,20000):
	x = dt* i
	cosx = (x2 - x1*cos(dt))/sin(dt)
	x3 = x1*cos(2*dt)+sin(2*dt)*cosx
	x1 = x2
	x2 = x3
	print(sin(x)-x3)

```

**Comment by wiz21b** [comment]
Here are the maths. Code will follow later. Please cite if you use them (I'll be happy!). Feel free to comment, correct, etc. I'll be very happy to collaborate ! Stefan.
[accurapple_speaker.pdf](https://github.com/AppleWin/AppleWin/files/12370274/accurapple_speaker.pdf)

**Comment by tomcw** [comment]
Thanks Stefan for sharing this work - it looks like a very detailed study, and some in depth analysis in deriving the equations from a good applied maths model.

**Comment by wiz21b** [comment]
Just to let you know I have now a prototype to compute the speaker source signal with 2 multiplications and 1 sum  per cycle (after computing constants and the first two cycles). And it looks stable (thanks to the exponential :-) ). That's twice as fast as Kris Kenaway's routine I think. I'll publish a new version of the maths soon. I have some prototype but I'd like to make a full version first before publishing the code; just to ensure I didn't miss anything). However, I am more and more concerned about the general "oscillator" idea. I'm not sure that's the best approach... I'll have to think about it (and make hundreds of tests).

For the record, I've been working on this for a bit less than a year now... I've made dozens of tests, recordings, analysis, etc.

**Comment by wiz21b** [comment]
I have updated my maths document. It's now located [here (gitlab)](https://gitlab.com/wiz21/accurapple/-/blob/main/Documentations/maths.pdf)

**Comment by tomcw** [comment]
@univta0001 - I've created a [branch](https://github.com/AppleWin/AppleWin/tree/GH1088-spkr-univta0001) with your modified Speaker.cpp (ie. implementation of the Kansasfest 2022 algorithm) from August 3rd ([above](https://github.com/AppleWin/AppleWin/issues/1088#issuecomment-1664008380)).

As an initial check, this code's speaker output is now silent in AppleWin:
```
300: LDA $C030
303: JMP 300
```
Then trying A2Stream's Sandstorm, I've included 2 .mp3's in [sandstorm_(univta001).zip](https://github.com/AppleWin/AppleWin/files/12386198/sandstorm_.univta001.zip):

1. sandstorm_(univta001)-emu6502.mp3
   - just an mp3 of your "Emu6502 using Kansasfest 2022 algorithm" in emu6502_darude_44100_stereo_32float.zip, from March 14th ([above](https://github.com/AppleWin/AppleWin/issues/1088#issuecomment-1468193905))
2. sandstorm_(univta001)-AppleWin.mp3
   - an mp3 of the output from AppleWin using your modified Speaker.cpp to implement the same "Kansasfest 2022 algorithm"

Note that the initial 0.5s of lead-in should be silent, emu6502 is a much higher frequency/lower volume (so you have to turn the volume up to hear it), whereas for AppleWin it's a lower frequency "buzz" which is audible at a normal volume setting.

Since it's the same algorithm then why does AppleWin have this lower frequency "buzz" for silent parts?
Hmm... it might be due to how AppleWin does the 1MHz -> 44.1kHz downsampling.

**Comment by oliverschmidt** [comment]
> Note that the initial 0.5s of lead-in should be silent, emu6502 is a much higher frequency/lower volume (so you have to turn the volume up to hear it), whereas for AppleWin it's a lower frequency "buzz" which is audible at a normal volume setting.

At least the headphone output of my IIgs doesn't show any of both. I'm pretty sure about that because about every cycle counting bug in A2Stream led to such artifacts.

**Comment by univta0001** [comment]
The higher frequency is probably due to the 48000 sampling frequency used in emu6502 vs 44100 used in AppleWin. For the constant used in code in [August 3](https://github.com/AppleWin/AppleWin/issues/1088#issuecomment-1664008380) for c1 and c2. It is obtained using the following snippet from python code where freq variable is the resonance frequency in the damped harmonics of the speaker. The sample_rate is computed as 17030 * 60 Hz (which is approximately the Apple 2 speed)

```python
            import numpy as np

            sample_rate = 1021800
            damping = -2000
            freq = 3880
            dt = np.float64(1 / sample_rate)
            w = np.float64(freq * 2 * np.pi * dt)
            d = damping * dt
            e = np.exp(d)
            c1 = np.float32(2 * e * np.cos(w))
            c2 = np.float32(e * e)
```

**Comment by wiz21b** [comment]
Somehow I missed a simplification in the maths. So I updated my [maths.pdf](https://gitlab.com/wiz21/accurapple/-/blob/main/Documentations/maths.pdf) so it reflects the code in the comment above. It will help those who want to know how one gets to a simple sum of product (IMHO guessing that just by reading some python code is pretty hard).

**Comment by univta0001** [comment]
> @univta0001 - I've created a [branch](https://github.com/AppleWin/AppleWin/tree/GH1088-spkr-univta0001) with your modified Speaker.cpp (ie. implementation of the Kansasfest 2022 algorithm) from August 3rd ([above](https://github.com/AppleWin/AppleWin/issues/1088#issuecomment-1664008380)).
> 
> As an initial check, this code's speaker output is now silent in AppleWin:
> 
> ```
> 300: LDA $C030
> 303: JMP 300
> ```
> 
> Then trying A2Stream's Sandstorm, I've included 2 .mp3's in [sandstorm_(univta001).zip](https://github.com/AppleWin/AppleWin/files/12386198/sandstorm_.univta001.zip):
> 
>     1. sandstorm_(univta001)-emu6502.mp3
>        
>        * just an mp3 of your "Emu6502 using Kansasfest 2022 algorithm" in emu6502_darude_44100_stereo_32float.zip, from March 14th ([above](https://github.com/AppleWin/AppleWin/issues/1088#issuecomment-1468193905))
> 
>     2. sandstorm_(univta001)-AppleWin.mp3
>        
>        * an mp3 of the output from AppleWin using your modified Speaker.cpp to implement the same "Kansasfest 2022 algorithm"
> 
> 
> Note that the initial 0.5s of lead-in should be silent, emu6502 is a much higher frequency/lower volume (so you have to turn the volume up to hear it), whereas for AppleWin it's a lower frequency "buzz" which is audible at a normal volume setting.
> 
> Since it's the same algorithm then why does AppleWin have this lower frequency "buzz" for silent parts? Hmm... it might be due to how AppleWin does the 1MHz -> 44.1kHz downsampling.

Recently i have make the emu6502 public at https://github.com/univta0001/emu6502. You can use it to make a comparison by running a2stream with applewin. The implementation of the emu6502 audio is at https://github.com/univta0001/emu6502/blob/main/emulator/src/audio.rs.
